posts/a-guinea-pig-in-the-cloud.md

```diff
@@ -0,0 +1,389 @@
+
+<table>
+  <tr>
+   <td><strong>Method</strong>
+   </td>
+   <td><strong>HTTP Verb</strong>
+   </td>
+   <td><strong>Request example \
+(for object Blog and id 123)</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>findRecord
+   </td>
+   <td>GET
+   </td>
+   <td>/blogs/123
+   </td>
+  </tr>
+  <tr>
+   <td>findRecord
+   </td>
+   <td>GET
+   </td>
+   <td>/blogs
+   </td>
+  </tr>
+  <tr>
+   <td>createRecord
+   </td>
+   <td>POST
+   </td>
+   <td>/blogs
+   </td>
+  </tr>
+  <tr>
+   <td>push
+   </td>
+   <td>PUT
+   </td>
+   <td>/blogs/123
+   </td>
+  </tr>
+  <tr>
+   <td>deleteRecord
+   </td>
+   <td>DELETE
+   </td>
+   <td>/blogs/123
+   </td>
+  </tr>
+</table>
+
+
+To be more accurate, the default implementation of Ember Data is based on the [JSON API Specification](https://jsonapi.org/). This specification clearly details both the different endpoints and the content of the messages that are generated, especially for the relationships between objects. 
+
+A JSON API compliant back-end should implement the HTTP vocabulary as depicted below :
+
+![JSON API](https://github.com/bpinel/blog-usa/blob/master/images/2019/02/API-Gateway.png)
+
+To achieve the implementation of this backend, we will rely mainly on three AWS services : 
+
+*   A new API declared in the **API Gateway** that support the awaited REST semantic,
+*   A Lambda function written on Javascript and running on Node.js 8.1, systematically called by the API and in charge of the logical processing of incoming parameters and also of the handling of potential relationships
+*   A DynamoDB table, in charge of persisting all the processed objects independently of their content.
+
+This main infrastructure for the backend is completed by the following elements:
+
+
+
+*   Three **S3 buckets**:
+    *   The first one (_lambda-jsonapi-code-bucket_) is simply used for provisioning the lambda code at its creation time,
+    *   The other two (_ember-aws-ehipster-staging_ et _ember-aws-ehipster-production_) will be used later to store the 'staging' and 'production' versions of the application.  
+*   An **IAM role** (_lambda_jsonapi_) allowing the lambda function to get access to the DynamoDB database,
+*   Two entries in **CloudFront** corresponding respectively to the buckets of 'staging' and 'production'
+
+
+### Building the backend infrastructure in AWS
+
+Rather than going through a manual construction of this infrastructure thanks to the AWS Web console, I have privilege an automated approach based on [terraform](https://www.terraform.io/) scripts, largely easier to read than their [CloudFormation](https://aws.amazon.com/fr/cloudformation/) counterparts. All necessary files are available in the cloud directory of the gitlab project: [https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud](https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud) that you should get on your local computer.
```

Comment:
This first sentence is phrased awkwardly.

Reply:
Tell me if the following sentences are clearer. 
```suggestion
Rather than going through a step-by-step construction of this infrastructure thanks to the AWS Web console, I have privilege an automated approach based on [terraform](https://www.terraform.io/) scripts. To my mind, Terraform scripts are definitely easier to write and to maintain than their [CloudFormation](https://aws.amazon.com/fr/cloudformation/) counterparts. All necessary files are available in the cloud directory of the gitlab project: [https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud](https://gitlab.ippon.fr/bpinel/ember-aws-ehipster/tree/master/cloud) that you should download on your local computer.
```